File: frontend-preview/src/lib/api/health.ts
/**
 * Health & Retry Queue API
 * Provides health monitoring for external services and retry queue management
 */

import { apiClient } from './client';

// ─── Health Check Types ───────────────────────────────────────────

export interface ServiceStatus {
	reachable: boolean;
	message: string;
}

export interface HealthStatus {
	healthy: boolean;
	qdrant: ServiceStatus;
	bm25: ServiceStatus;
	embedding: ServiceStatus;
}

export interface QdrantDiagnostic {
	reachable: boolean;
	version: string | null;
	collection_exists: boolean;
	points_count: number;
	error: string | null;
}

export interface QdrantHealthStatus {
	healthy: boolean;
	circuit_breaker: string;
	diagnostic: QdrantDiagnostic;
}

export interface EmbeddingHealthStatus {
	healthy: boolean;
	model_name: string | null;
	message: string;
}

export interface Bm25HealthStatus {
	enabled: boolean;
	connected: boolean;
	index_path: string | null;
}

// ─── Retry Queue Types ────────────────────────────────────────────

export interface RetryQueueStatus {
	pending_count: number;
	is_empty: boolean;
}

export interface RetryQueueProcessResponse {
	processed: number;
	message: string;
}

export interface RetryQueueClearResponse {
	cleared: number;
	message: string;
}

// ─── API ──────────────────────────────────────────────────────────

export const healthApi = {
	// Unified health check
	getHealth: () => apiClient.get<HealthStatus>('/api/health'),

	// Qdrant detailed diagnostics
	getQdrantHealth: () => apiClient.get<QdrantHealthStatus>('/api/health/qdrant'),

	// Embedding service health
	getEmbeddingHealth: () => apiClient.get<EmbeddingHealthStatus>('/api/health/embedding'),

	// BM25 index health
	getBm25Health: () => apiClient.get<Bm25HealthStatus>('/api/health/bm25'),

	// Retry queue status
	getRetryQueueStatus: () => apiClient.get<RetryQueueStatus>('/api/retry-queue'),

	// Manually trigger retry queue processing
	processRetryQueue: () =>
		apiClient.post<RetryQueueProcessResponse>('/api/retry-queue/process', {}),

	// Clear retry queue
	clearRetryQueue: () =>
		apiClient.delete<RetryQueueClearResponse>('/api/retry-queue'),
};

File: crates/app/cce-api/src/models/health.rs
//! Health check models

use serde::{Deserialize, Serialize};

/// Health status response
#[derive(Debug, Serialize, Deserialize)]
pub struct HealthStatus {
    /// Overall health
    pub healthy: bool,
    /// Qdrant status
    pub qdrant: ServiceStatus,
    /// BM25 status
    pub bm25: ServiceStatus,
    /// Embedding status
    pub embedding: ServiceStatus,
}

/// Service status
#[derive(Debug, Serialize, Deserialize)]
pub struct ServiceStatus {
    /// Whether the service is reachable
    pub reachable: bool,
    /// Human-readable message
    pub message: String,
}

/// Qdrant health response
#[derive(Debug, Serialize, Deserialize)]
pub struct QdrantHealthResponse {
    pub healthy: bool,
    pub circuit_breaker: String,
    pub diagnostic: QdrantDiagnostic,
}

/// Qdrant diagnostic
#[derive(Debug, Serialize, Deserialize)]
pub struct QdrantDiagnostic {
    #[serde(default)]
    pub reachable: bool,
    #[serde(default)]
    pub version: Option<String>,
    #[serde(default)]
    pub collection_exists: bool,
    #[serde(default)]
    pub points_count: u64,
    #[serde(default)]
    pub error: Option<String>,
}

/// Embedding health response
#[derive(Debug, Serialize, Deserialize)]
pub struct EmbeddingHealthResponse {
    pub healthy: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model_name: Option<String>,
    pub message: String,
}

/// BM25 health response
#[derive(Debug, Serialize, Deserialize)]
pub struct Bm25HealthResponse {
    pub enabled: bool,
    pub connected: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub index_path: Option<String>,
}

/// Retry queue status response
#[derive(Debug, Serialize, Deserialize)]
pub struct RetryQueueStatusResponse {
    pub pending_count: usize,
    pub is_empty: bool,
}

/// Retry queue process response
#[derive(Debug, Serialize, Deserialize)]
pub struct RetryQueueProcessResponse {
    pub processed: usize,
    pub message: String,
}

/// Retry queue clear response
#[derive(Debug, Serialize, Deserialize)]
pub struct RetryQueueClearResponse {
    pub cleared: usize,
    pub message: String,
}
